torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
+  auto val = [](const Tensor& t) {return t.item<double>(); };
+
+  auto& group = param_groups_.at(0);
+  auto& _params = group.params();
+  const auto& options = static_cast<const LBFGSOptions&>(group.options());
+  auto lr = options.lr();
+  auto max_iter = options.max_iter();
+  auto max_eval = options.max_eval();
+  auto tolerance_grad = options.tolerance_grad();
+  auto tolerance_change = options.tolerance_change();
+  auto line_search_fn = options.line_search_fn();
+  auto history_size = options.history_size();
+
+  // NOTE: LBFGS has only global state, but we register it as state for
+  // the first param, because this helps with casting in load_state_dict
+  auto param_state = state_.find(c10::guts::to_string(_params.at(0).unsafeGetTensorImpl()));
+  if (param_state == state_.end()) {
+    state_[c10::guts::to_string(_params.at(0).unsafeGetTensorImpl())] = std::make_unique<LBFGSParamState>();
+  }
+  auto& state = static_cast<LBFGSParamState&>(*state_[c10::guts::to_string(_params.at(0).unsafeGetTensorImpl())]);
+  // evaluate initial f(x) and df/dx
+  Tensor orig_loss;
+  {
+    torch::AutoGradMode enable_grad(true);
+    orig_loss = closure();
+  }
+
+  auto loss = orig_loss.clone(at::MemoryFormat::Contiguous);
+  auto current_evals = 1;
+  state.func_evals(state.func_evals() + 1);
+  auto flat_grad = _gather_flat_grad();
+  auto opt_cond = (val(flat_grad.abs().max()) <= tolerance_grad);
+
+  // optimal condition
+  if (opt_cond) {
+    return orig_loss;
+  }
+
+  // tensors cached in state (for tracing)
+  auto& d = state.d();
+  auto& t = state.t();
+  auto& old_dirs = state.old_dirs();
+  auto& old_stps = state.old_stps();
+  auto& ro = state.ro();
+  auto& H_diag = state.H_diag();
+  auto& prev_flat_grad = state.prev_flat_grad();
+  auto& prev_loss = state.prev_loss();
+  auto& al = state.al();
+
+  int n_iter = 0;
   Tensor ONE = torch::tensor(1, flat_grad.options());
 
-  int64_t n_iter = 0;
-  while (n_iter < options.max_iter()) {
-    n_iter++;
-    state_n_iter++;
+  // optimize for a max of max_iter iterations
+  while(n_iter < max_iter) {
+    // keep track of nb of iterations
+    n_iter += 1;
+    state.n_iter(state.n_iter() + 1);
 
-    if (state_n_iter == 1) {
+    // compute gradient descent direction
+    if (state.n_iter() == 1) {
       d = flat_grad.neg();
       H_diag = ONE;
-      prev_flat_grad = flat_grad.clone(at::MemoryFormat::Contiguous);
     } else {
-      Tensor y = flat_grad.sub(prev_flat_grad);
-      Tensor s = d.mul(t);
-      Tensor ys = y.dot(s);
-
-      if (ys.item<float>() > 1e-10) {
+      // do lbfgs update (update memory)
+      auto y = flat_grad.sub(prev_flat_grad);
+      auto s = d.mul(t);
+      auto ys = y.dot(s); // y*s
+      if (val(ys) > 1e-10) {
         // updating memory
-
-        if (old_dirs.size() == options.history_size()) {
-          // shift history by one (limited memory)
+        if (int(old_dirs.size()) == history_size) {
```

Comment:
`history_size` is `int64_t` type, so maybe 
```suggestion
        if (static_cast<int64_t>(old_dirs.size()) == history_size) {
```